homeassistant/components/octoprint.py

```diff
@@ -4,57 +4,115 @@
 
 _LOGGER = logging.getLogger(__name__)
 
 DOMAIN = 'octoprint'
+ENTITY_ID_FORMAT = 'octoprint.{}'
 CONF_NUMBER_OF_TOOLS = 'number_of_tools'
 CONF_BED = 'bed'
 
+# pylint: disable=invalid-name
+T = TypeVar('T')
+# pylint: enable=invalid-name
+
+
+# This version of ensure_list interprets an empty dict as no value
+def ensure_list(value: Union[T, Sequence[T]]) -> Sequence[T]:
+    """Wrap value in list if it is not one."""
+    if value is None or (isinstance(value, dict) and not value):
+        return []
+    return value if isinstance(value, list) else [value]
+
+
 CONFIG_SCHEMA = vol.Schema({
-    DOMAIN: vol.Schema({
-        vol.Required(CONF_API_KEY): cv.string,
+    DOMAIN: vol.All(ensure_list, [vol.Schema({
+        vol.Required(CONF_NAME): cv.string,
         vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT): cv.string,
+        vol.Optional(CONF_SSL, default=False): cv.boolean,
+        vol.Required(CONF_API_KEY): cv.string,
         vol.Optional(CONF_NUMBER_OF_TOOLS, default=0): cv.positive_int,
         vol.Optional(CONF_BED, default=False): cv.boolean
-    }),
+    })])
 }, extra=vol.ALLOW_EXTRA)
 
 
-def setup(hass, config):
+@asyncio.coroutine
+def async_setup(hass, config):
     """Set up the OctoPrint component."""
-    base_url = 'http://{}/api/'.format(config[DOMAIN][CONF_HOST])
-    api_key = config[DOMAIN][CONF_API_KEY]
-    number_of_tools = config[DOMAIN][CONF_NUMBER_OF_TOOLS]
-    bed = config[DOMAIN][CONF_BED]
-
-    hass.data[DOMAIN] = {"api": None}
-
-    try:
-        octoprint_api = OctoPrintAPI(base_url, api_key, bed, number_of_tools)
-        hass.data[DOMAIN]["api"] = octoprint_api
-        octoprint_api.get('printer')
-        octoprint_api.get('job')
-    except requests.exceptions.RequestException as conn_err:
-        _LOGGER.error("Error setting up OctoPrint API: %r", conn_err)
+    octoprints = hass.data[DOMAIN] = {}
+
+    @asyncio.coroutine
+    def async_setup_octoprint(octo_config):
+        """Set up an Octoprint service."""
+        name = octo_config[CONF_NAME]
+        _LOGGER.debug("Configuring Octoprint %s", octo_config[CONF_NAME])
+
+        schema = 'http'
+        if CONF_SSL in octo_config:
+            if octo_config[CONF_SSL] is True:
+                schema = 'https'
+
+        if CONF_PORT in octo_config:
+            url = '{}:{}'.format(
+                octo_config[CONF_HOST],
+                octo_config[CONF_PORT]
+            )
+        else:
+            url = '{}'.format(octo_config[CONF_HOST])
+
+        base_url = '{}://{}/api/'.format(schema, url)
+        api_key = octo_config[CONF_API_KEY]
+        number_of_tools = octo_config[CONF_NUMBER_OF_TOOLS]
+        bed = octo_config[CONF_BED]
 
+        try:
+            octoprint_api = OctoPrintAPI(
+                name,
+                base_url,
+                api_key,
+                bed,
+                number_of_tools
+            )
+            octoprint_api.get('printer')
```

Comment:
This call is using the requests library, ie sync I/O. We can't do that in a coroutine.